src/ui.js

```diff
@@ -125,21 +112,70 @@ function handleDragOver(evt) {
 
 
 function buildUploadModal(numFiles) {
-    let getModalContent = numLoaded => `<h1>Reading GPX files...</h1>
-<span id="">${numLoaded} loaded of <b>${numFiles}</b>`;
+    let numLoaded = 0;
+    let failures = [];
+    let getModalContent = () => `
+        <h1>Reading GPX files...</h1>
+        <p>${numLoaded} loaded${
+            failures.length ? `, <span class='failures'>${failures.length} failed</span>` : ``
+        } of <b>${numFiles}</b></p>`;
 
     let modal = picoModal({
-        content: getModalContent(0),
-        closeButton: false,
+        content: getModalContent(),
         escCloses: false,
         overlayClose: false,
         overlayStyles: styles => {
             styles.opacity = 0.1;
         },
-    })
+    });
+
+    modal.afterCreate(() => {
+        // Do not allow the modal to be closed before loading is complete.
+        // PicoModal does not allow for native toggling
+        modal.closeElem().style.display = 'none';
+    });
+
+    modal.afterClose(() => modal.destroy());
 
-    modal.progress = loaded => {
-        modal.modalElem().innerHTML = getModalContent(loaded);
+    // Override the content of the modal, without removing the close button.
+    // PicoModal does not allow for native content overwriting.
+    modal.setContent = body => {
+        Array.from(modal.modalElem().childNodes).forEach(child => {
+            if (child != modal.closeElem()) {
+                modal.modalElem().removeChild(child);
+            }
+        });
+        modal.modalElem().insertAdjacentHTML('afterbegin',body);
+    };
+
+    modal.addFailure = failure => {
+        failures.push(failure);
+        modal.setContent(getModalContent());
+    };
+
+    modal.addSuccess = () => {
+        numLoaded++;
+        modal.setContent(getModalContent());
+    };
+
+    modal.finished = () => {
+        if (failures.length == 0)
+            return modal.close();
+        let failedItems = failures.map(failure => `<li>${failure.name}</li>`);
+        modal.setContent(`
+            <h1>GPX files loaded</h1>
+            <p>
+                Loaded ${numLoaded}, 
+                <span class="failures">
+                    ${failures.length} failure${failures.length==1?'':'s'}:
+                </span>
+            </p>
+            <ul class="failures">${failedItems.join('')}</ul>`);
+        modal.closeElem().style.display = '';
+        modal.options({
+            escCloses: true,
+            overlayClose: true,
+        });
```

Comment:
When the failed items are shown, this view can be dismissed with esc/×/click outside. All the failed files are listed

Reply:
Could this also be added as a code comment?

Reply:
443e466e3513891ef4bff9c5029d4e63c64d09ea